File: src/main/java/com/jeroensteenbeeke/topiroll/beholder/entities/Portrait.java
/*
 * This file is part of Beholder
 * Copyright (C) 2016 - 2023 Jeroen Steenbeeke
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.jeroensteenbeeke.topiroll.beholder.entities;

import com.jeroensteenbeeke.hyperion.data.BaseDomainObject;
import com.jeroensteenbeeke.hyperion.webcomponents.entitypage.DefaultFieldType;
import com.jeroensteenbeeke.hyperion.webcomponents.entitypage.annotation.EntityFormField;

import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import jakarta.persistence.*;
import java.io.Serializable;
import java.sql.Blob;
import java.util.ArrayList;
import java.util.List;

@Entity
public class Portrait extends BaseDomainObject implements AmazonStored {

	private static final long serialVersionUID = 1L;
	@Id
	@SequenceGenerator(sequenceName = "SEQ_ID_Portrait", name = "Portrait", initialValue = 1, allocationSize = 1)
	@GeneratedValue(generator = "Portrait", strategy = GenerationType.SEQUENCE)
	@Access(value = AccessType.PROPERTY)
	private Long id;

	@Column(nullable = false)
	@EntityFormField(label = "Name", required = true)
	private String name;

	@ManyToOne(fetch = FetchType.LAZY, optional = true)
	@JoinColumn(name = "campaign")
	@EntityFormField(label = "Campaign", type = DefaultFieldType.DropDownChoice.class)
	private Campaign campaign;

	@Column(name = "amazon_key")
	private String amazonKey;

	@OneToMany(mappedBy = "portrait", fetch = FetchType.LAZY)
	private List<PortraitVisibility> visibilities = new ArrayList<PortraitVisibility>();

	@ManyToOne(fetch = FetchType.LAZY, optional = false)
	@JoinColumn(name = "owner")
	private BeholderUser owner;

	public Long getId() {
		return id;
	}

	public void setId(@NotNull Long id) {
		this.id = id;
	}

	@Override
	public final Serializable getDomainObjectId() {
		return getId();
	}

	@NotNull
	public BeholderUser getOwner() {
		return owner;
	}

	public void setOwner(@NotNull BeholderUser owner) {
		this.owner = owner;
	}

	@NotNull
	public String getName() {
		return name;
	}

	public void setName(@NotNull String name) {
		this.name = name;
	}

	@NotNull
	public List<PortraitVisibility> getVisibilities() {
		return visibilities;
	}

	public void setVisibilities(
		@NotNull List<PortraitVisibility> visibilities) {
		this.visibilities = visibilities;
	}

	@Nullable
	@Override
	public String getAmazonKey() {
		return amazonKey;
	}

	public void setAmazonKey(@Nullable String amazonKey) {
		this.amazonKey = amazonKey;
	}

	@Nullable
	public Campaign getCampaign() {
		return campaign;
	}

	public void setCampaign(@Nullable Campaign campaign) {
		this.campaign = campaign;
	}

}

File: src/main/java/com/jeroensteenbeeke/topiroll/beholder/web/components/js/portrait.js
/*
 * This file is part of Beholder
 * (C) 2017 Jeroen Steenbeeke
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
function removePortraitImages(containerId) {
    var container = document.getElementById(containerId);
    var i;
    var remove = [];

    for (i = 0; i < container.childNodes.length; i++) {
        var elem = container.childNodes.item(i);
        if (elem.tagName === 'IMG') {
            remove.push(elem);
        }
    }

    remove.forEach(function(e) {
        container.removeChild(e);
    });


}

function updatePortraits(data) {
    var portraits = data.portraits;

    // First, remove all existing portraits
    removePortraitImages('fs');
    removePortraitImages('left');
    removePortraitImages('right');
    removePortraitImages('topleft');
    removePortraitImages('topright');
    removePortraitImages('bottomleft');
    removePortraitImages('bottomright');

    portraits.forEach(function(p) {
        var target = '';

        if (p.location === 'full') {
            target = 'fs';
        } else if (p.location === 'left') {
            target = 'left';
        } else if (p.location === 'right') {
            target = 'right';
        } else if (p.location === 'top_left') {
            target = 'topleft';
        } else if (p.location === 'top_right') {
            target = 'topright';
        } else if (p.location === 'bottom_left') {
            target = 'bottomleft';
        } else if (p.location === 'bottom_right') {
            target = 'bottomright';
        }

        if (target !== '') {
            Images.load(p.url, function(img) {
                document.getElementById(target).appendChild(img);
            });
        }
    });
}
